Mocha qe baseline

PastelBelem8 opened the pull request:
Goal: Implement MOCHA reader and Roberta-base regression model

Main changes
- Add `training_config/mocha_eqqa.jsonnet`: Configuration file for the MOCHA-ROBERTA QE experiment.
- Add `quality_estimation/mocha_eqqa_reader.py`: MOCHA dataset reader. This considers the preprocessed version of the [files available here]([here](https://drive.google.com/file/d/1tUyjF7QX3ZAozy0-uhyhtpegzWGt-diU/view?usp=sharing)).
- Add `quality_estimation/mocha_eqqa_roberta_model.py`: a roberta-base transformer with a regression layer on top.

PastelBelem8 committed:
Update gitignore

PastelBelem8 committed:
Add README: add setup instructions

PastelBelem8 committed:
Add MOCHA preprocessing notebook

PastelBelem8 committed:
Add `split_data` to MOCHA preprocessor

PastelBelem8 committed:
Implement mocha_eqqa_reader._read

In order to create the baseline for QE, we need to specify the reader
class.

PastelBelem8 committed:
Fix typo

PastelBelem8 committed:
Add initial training_config/mocha-qe-baseline config

This commit marks a working version of the mocha reader.
The instances were both successfully loaded. You can download the
dataset from the following
[link](https://drive.google.com/file/d/1tUyjF7QX3ZAozy0-uhyhtpegzWGt-diU/view?usp=sharing)

PastelBelem8 committed:
Add working version of MOCHA_EQQA_ROBERTA model

PastelBelem8 committed:
Update README.md w/ running command

PastelBelem8 committed:
Update MOCHA-EQQA-ROBERTA: comment PEARSON measure

Pearson correlation metric is giving an error

PastelBelem8 committed:
Update MOCHA EQQA models: Add min-max scaling to target

PastelBelem8 committed:
Update MOCHA-READER: Add checks for target values, remove imports

PastelBelem8 committed:
Update MOCHA-ROBERTA model; Add 'train_base', and init layers

PastelBelem8 committed:
Update MOCHA-READER: Fix typo

PastelBelem8 committed:
Update MOCHA config

PastelBelem8 committed:
Update MOCHA-EQQA-PREPROCESSING notebook

pdasigi commented on quality_estimation/mocha_eqqa_reader.py:
```diff
@@ -0,0 +1,164 @@
+            if use_context else []
+                
+
+        if use_context and len(tokenized_context) > allowed_context_length:
```
This would never be triggered, right, because you are truncating to `tokenized_context` to `allowed_context_length` above?

PastelBelem8 replied:
You're right! I've updated [this](https://github.com/pdasigi/eqqa/commit/af6d012dee8bd3c23e75d8aa00c90d944ed9c302#diff-8100ca80c5c8be639e12520d4912c1238843038deb6717ba1ef1ebd98341f392R139) to ensure that we get proper logging of truncated instances.

Additionally, while this is not yet a concern, it might happen that either answers are too large or the question is too large... To accommodate for that, I added a `truncated_question` parameter to differentiate between the two cases.

pdasigi commented on quality_estimation/mocha_eqqa_reader.py:
```diff
@@ -0,0 +1,164 @@
+            + tokenized_context
+        )
+
+        if len(answer_question) < self.max_document_length:
```
You don't need this if-else block, and you also don't need to include an `attention_mask` in `fields`. I did this for the MLP model to ensure that all the inputs are of the same length, which was needed because the MLP's input dimensionality was fixed. Since you're passing the input to a RoBERTa model, you don't have to pad the inputs here. AllenNLP will take care of padding each batch appropriately internally. This should also address the "FIXME" you have in the model code.

PastelBelem8 replied:
What exactly makes RoBERTa different than MLP ? Why can't AllenNLP do proper padding for MLP  too? Doesn't RoBERTa also have a fixed dimension (?)

pdasigi commented on quality_estimation/mocha_eqqa_roberta_model.py:
```diff
@@ -0,0 +1,132 @@
+        **kwargs
+    ):
+        super().__init__(vocab, **kwargs)
+        config = AutoConfig.from_pretrained(transformer_model_name)
```
You don't need to get the config if you are not making changes to it. You can call `AutoModel.from_pretrained()` below without passing in the `config` and it will have the same effect.

pdasigi commented on quality_estimation/mocha_eqqa_roberta_model.py:
```diff
@@ -0,0 +1,132 @@
+        config = AutoConfig.from_pretrained(transformer_model_name)
+        self.transformer = AutoModel.from_pretrained(transformer_model_name, config=config)
+
+        # We do not want to train the base transformer models
```
When do we ever want to do this? If you want to keep it, maybe set the default value of `train_base` to `True`, so that you won't freeze the base model accidentally.

pdasigi commented on quality_estimation/mocha_eqqa_roberta_model.py:
```diff
@@ -0,0 +1,132 @@
+        # Define **Regression** head
+        # ----------------------------------------------------------------
+        self.regression_layer = [
+            torch.nn.Linear(self.embedding_dim, hidden_size),
```
Shouldn't this be `self.tranformer.config.hidden_size` instead of the dimensionality of the word embeddings (`self.embedding_dim`)? They're probably the same value but it is the output of the transformer that you're feeding into the MLP, right?

PastelBelem8 replied:
Yes, it is! I wasn't sure if the `self.transformer.config.hidden_size` parameter would be the same throughout all layers. 


I've just confirmed this is the same value as I was using, so I'm updating it.

pdasigi commented on quality_estimation/mocha_eqqa_roberta_model.py:
```diff
@@ -0,0 +1,132 @@
+        input_ids = util.get_token_ids_from_text_field_tensors(input_text)
+        
+        # FIXME 
+        # - Why do we overwrite our attention mask? 
```
See comment in the dataset reader above.

pdasigi commented on training_config/mocha_eqqa.jsonnet:
```diff
@@ -0,0 +1,59 @@
+      "num_gradient_accumulation_steps": num_gradient_accumulation_steps,
+      "patience": epochs,
+      "enable_default_callbacks": false,
+      "use_amp": true,
```
This argument enables mixed precision training (FP16). It is better to start with full precision (FP32) and turn this on only if you're running out of memory on your GPU. So maybe just set it to `false` to begin with.

pdasigi commented on quality_estimation/mocha_eqqa_roberta_model.py:
```diff
@@ -0,0 +1,132 @@
+
+@Model.register("mocha_eqqa_roberta")
+class MochaQualityEstimator(Model):
+    def _init_layers(self):
```
I'm curious why you need this initialization. Do you think Pytorch's default initialization is not good?

PastelBelem8 replied:
Honestly, I wasn't that sure about the layers initialization. I've now realized that depending on the layer type, different initialization are performed. 

Since I was obtaining some different results and in my previous code (in Computer Vision) I used to initialize the layers myself, I decided to try my own normalization to see if the results improved.

pdasigi reviewed (approved):
Looks good! I left a bunch of comments that essentially say "You don't need to do this", but I think the code is generally fine.

PastelBelem8 committed:
Fix truncation check to guarantee proper logging

PastelBelem8 committed:
Remove attention mask from base model

In the PR !2, @pdasigi [comments](https://github.com/pdasigi/eqqa/pull/2#discussion_r862259908)
on the fact that the attention mask is not needed.
Effectively, we were not using in the model, hence this
commit simplifies the interface and affects both reader and model class.

PastelBelem8 committed:
Remove config option from mocha-eqqa-roberta model

PastelBelem8 committed:
Update `train_base` parameter

PastelBelem8 committed:
Update w/ standard configuration for hidden_size

PastelBelem8 committed:
Update config: set default 'use_amp' to false

PastelBelem8 committed:
Remove initialization of layers

PastelBelem8 committed:
Update running scripts
